wallet/src/hooks/User/index.jsx

```diff
@@ -105,58 +110,55 @@ export const UserProvider = ({ children }) => {
+        await parsedEvent.historicalData
           .sort((a, b) => a.block.blockNumberL2 - b.block.blockNumberL2)
           .reduce(async (acc, curr) => {
             await acc; // Acc is a promise so we await it before processing the next one;
             return blockProposedEventHandler(curr, [zkpPrivateKey], [nullifierKey]); // TODO Should be array
           }, Promise.resolve());
 
-        // We want to verify that the received block is from the current contract deployment.
-        // for this, we store the lastBlock received in lastBlock, and compare the hash with the previousBlockHash
-        // from the received block. If they don't match, then there has been a redeployment.
-        if (Number(parsed.maxBlock) !== 1) {
-          if (
-            parsed.historicalData[parsed.historicalData.length - 1].block.previousBlockHash !==
-              lastBlock.blockHash &&
-            Number(lastBlock.blockHash) !== 0
-          ) {
-            // resync
-            console.log('Resync DB');
-            emptyStoreBlocks();
-            emptyStoreTimber();
-            Storage.shieldAddressSet();
-          } else if (
-            parsed.historicalData[parsed.historicalData.length - 1].block.previousBlockHash ===
-              lastBlock.blockHash ||
-            Number(lastBlock.blockHash) === 0
-          ) {
-            setLastBlock(parsed.historicalData[parsed.historicalData.length - 1].block);
+        const indexDbBlockRoot = await getNthBlockRoot(parsedEvent.numberBlockLast);
+        const matchContract = indexDbBlockRoot === parsedEvent.rootBlockLast;
+        const isFirst = parsedEvent.numberBlockLast === null;
+
+        if (matchContract || isFirst) {
+          if (parsedEvent.isLast) {
+            console.log('Sync complete');
+            setState(previousState => {
+              return {
+                ...previousState,
+                chainSync: true,
+              };
+            });
+          } else {
+            console.log(
+              `Last local block number is ${parsedEvent.numberBlockNext}, emitting sync event to fetch next block.`,
+            );
             socket.send(
               JSON.stringify({
                 type: 'sync',
-                lastBlock:
-                  parsed.historicalData[parsed.historicalData.length - 1].block.blockNumberL2,
+                lastBlock: parsedEvent.numberBlockNext,
               }),
             );
           }
-        } else if (lastBlock.blockHash) {
-          console.log('Sync complete');
-          setState(previousState => {
-            return {
-              ...previousState,
-              chainSync: true,
-            };
-          });
+        } else {
+          console.log('Resync DB');
+          emptyStoreBlocks();
+          emptyStoreTimber();
+          Storage.shieldAddressSet();
         }
-      } else if (parsed.type === 'blockProposed') {
+      }
+
+      if (parsedEvent.type === 'blockProposed') {
```

Comment:
And exactly because these conditions are mutually exclusive, bare `if` should be preferred. The `else` if should be used only if you want to assert for an ordered set of conditions. In the future it is going to be refactored anyway (to something like you suggested above).

```js
const EVENT_HANDLERS_MAPPER = {
  sync: handleSync,
  blockProposed: handleBlockProposed,
  rollback: handleRollback,
};

const messageHandler = (event) => EVENT_HANDLERS_MAPPER[event.type](event);

```

Reply:
I understand what you say, from the point of view of performance, which in this case doesn't apply, doing many `if` impact negatively the throughput, mainly in this case with mutually-exclusive conditions.